Add configurable cross-day aggregation options

seyfahni opened the pull request:
This is my current draft for #4. Let's discuss about it!

Open topics:
* Naming is not consistent, I went with both `aggregation` and `cross-day handling` and mixed it all over the place. I'm not quite happy with either name.
* The need to add a new error just for this single case; err113 complains otherwise. Maybe that error should be more general and be used in all parsing methods?
* The large chunks of code added to the main, it might be a good idea to extract them elsewhere.

Please let me know what you think.

This PR closes #4.

aibor commented:
Thanks for the PR. I like the concept of configurable day sum strategies a lot.

>     * Naming is not consistent, I went with both `aggregation` and `cross-day handling` and mixed it all over the place. I'm not quite happy with either name.
> 
I'd drop the multi/cross day focus and go with something like `(day) aggregation strategy` or even `day sum strategy`, since `daySums` is used already anyway. The name for the default case (which is "ignore" in your draft) could be "single-day-only" or something like this?

>     * The need to add a new error just for this single case; err113 complains otherwise. Maybe that error should be more general and be used in all parsing methods?
Instead of adding an error type I'd suggest to add a generic sentinel error and use it like this, which makes the linter happy:
```
var ErrUnknownAggregationStrategy = errors.New("unknown aggregation strategy")
//...
return fmt.Errorf("%w: %s", ErrUnknownAggregationStrategy, strategyName)
```
This is only needed in the single resolving function (described later).

>     * The large chunks of code added to the main, it might be a good idea to extract them elsewhere.
I think moving all the aggregation handling to a dedicated file with dedicated tests for resolving names and correct aggregation behavior would be best.

On that note, I (and the linter) don't like the strategies being a global map. Instead, I'd prefer a function that resolves the name to the strategy struct and can be used when parsing the config. It returns the strategy struct or an error if the name is unknown. I'd use a switch case statement to implement it.

I'd store the strategy struct in the config struct directly. My idea of the config struct was that it provides the validated values ready to use. This way there's no need for another lookup (and invalid name handling) in the main function. In order to provide the name for debugging output the strategy struct could just get a name field.

What do yo think?

seyfahni commented:
I should've addressed all your points and tried to clean up my code a bit more as well. Let me know if there's anything left that you want to change.

Regarding that linter error, I'm not sure how to fix it:
```
Error: cmd/flextime/config.go:79:1: configRead returns generic interface (R) of type param any (ireturn)
  func configRead[R any](...) (R, error) { ... }
  ```

aibor commented:
> I should've addressed all your points and tried to clean up my code a bit more as well. Let me know if there's anything left that you want to change.

Cool, thanks. I'll review it this weekend.

> Regarding that linter error, I'm not sure how to fix it:
> 
> ```
> Error: cmd/flextime/config.go:79:1: configRead returns generic interface (R) of type param any (ireturn)
>   func configRead[R any](...) (R, error) { ... }
> ```
In my view, for this specific case it is perfectly fine to add a `//nolint:ireturn,nolintlint` tag.

aibor commented:
LGTM, thank you very much.
